# Torchvision Object Detection Training
## Author: Alex de la Paz (http://www.alexdelapaz.com)

Training using Torchvision `FasterRCNN_ResNet50` 2-stage model

A subset of the `xView dataset` is used
- http://xviewdataset.org/

This subset was selected to perform a training run that works in a reasonable amount of time on Colab GPUs or personal use GPUs

In [8]:
! git clone https://github.com/alexdelapaz/object-detection-torch

Cloning into 'object-detection-torch'...
remote: Enumerating objects: 1475, done.
remote: Counting objects: 100% (454/454), done.
remote: Compressing objects: 100% (399/399), done.
remote: Total 1475 (delta 92), reused 396 (delta 51), pack-reused 1021
Receiving objects: 100% (1475/1475), 341.03 MiB | 48.09 MiB/s, done.
Resolving deltas: 100% (336/336), done.


In [2]:
cd /content/object-detection-torch

/content/object-detection-torch


In [3]:
!git clone https://github.com/pytorch/vision.git

Cloning into 'vision'...
remote: Enumerating objects: 322836, done.
remote: Counting objects: 100% (821/821), done.
remote: Compressing objects: 100% (139/139), done.
remote: Total 322836 (delta 685), reused 804 (delta 675), pack-reused 322015
Receiving objects: 100% (322836/322836), 656.80 MiB | 35.00 MiB/s, done.
Resolving deltas: 100% (296708/296708), done.


# Force install `pycocotools` with bug fix for colab

In [4]:
!pip install cython
# Install pycocotools, the version by default in Colab
# has a bug fixed in https://github.com/cocodataset/cocoapi/pull/354
!pip install -U 'git+https://github.com/cocodataset/cocoapi.git#subdirectory=PythonAPI'

# Download TorchVision repo to use some files from
# references/detection
!git clone https://github.com/pytorch/vision.git
#!cd vision && git checkout v0.3.0
!cp vision/references/detection/utils.py ./
!cp vision/references/detection/transforms.py ./
!cp vision/references/detection/coco_eval.py ./
!cp vision/references/detection/engine.py ./
!cp vision/references/detection/coco_utils.py ./

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Cloning https://github.com/cocodataset/cocoapi.git to /tmp/pip-req-build-vrpkzpmb
  Running command git clone --filter=blob:none --quiet https://github.com/cocodataset/cocoapi.git /tmp/pip-req-build-vrpkzpmb
  Resolved https://github.com/cocodataset/cocoapi.git to commit 8c9bcc3cf640524c4c20a9c40e89cb6a2f2fa0e9
  Preparing metadata (setup.py) ... done
  Created wheel for pycocotools: filename=pycocotools-2.0-cp39-cp39-linux_x86_64.whl size=397990 sha256=c0f2e7614de81cdb2db42f4e7336f8c8faf151bf5293e7a98f6fd112aa8adb21
  Stored in directory: /tmp/pip-ephem-wheel-cache-a3nfbh_s/wheels/13/c1/d6/a321055f7089f1a6af654fbf794536b196999f082a9cb68a37
Successfully built pycocotools
  Attempting uninstall: pycocotools
    Found existing installation: pycocotools 2.0.6
    Uninstalling pycocotools-2.0.

## Train FasterRCNN_ResNet50
- 336 Training samples
- 84 Test samples
- 20 Epochs
- Save loss, eval, and model weights every 5 epochs
- Save weights every epoch
- `SGD` optimizer
- Learning rate of 0.0001

In [5]:
! python3 strategy_real.py -d 'artifacts' -n 'planes_sgd' -bs 4 -e 20 -s 5 -o sgd -lr 0.0001

Downloading: "https://download.pytorch.org/models/fasterrcnn_resnet50_fpn_coco-258fb6c6.pth" to /root/.cache/torch/hub/checkpoints/fasterrcnn_resnet50_fpn_coco-258fb6c6.pth
100% 160M/160M [00:02<00:00, 56.4MB/s]
------------------------------------------------------
Training: FasterRCNN_ResNet50

Saving planes_sgd artifacts to: artifacts/planes_sgd
------------------------------------------------------
Data [Total count]:	420
Data [Train count]:	336
Data [Test count] :	84

Training [epoch 1]:	out of 20
Training: 100% 84/84 [01:08<00:00,  1.22it/s]
Validation: 100% 84/84 [00:08<00:00, 10.43it/s]
Lowest train loss: 1.0867688655853271
Lowest val loss: 0.6924095153808594

Training [epoch 2]:	out of 20
Training: 100% 84/84 [01:04<00:00,  1.29it/s]
Validation: 100% 84/84 [00:07<00:00, 10.71it/s]
Lowest train loss: 0.6528642177581787
Lowest val loss: 0.5605618357658386

Training [epoch 3]:	out of 20
Training: 100% 84/84 [01:04<00:00,  1.29it/s]
Validation: 100% 84/84 [00:08<00:00,  9.96it/s]


## Run analysis
- display training, validation losses
- display minimum training and validation losses
- display COCO eval data
- save loss graph in `artifacts/` under `/project_name/`

In [6]:
!python3 analysis.py -d 'artifacts' -n 'planes_sgd'


-------------------------------------------------------------------------------
Analysis for |FasterRCNN_ResNet50: 20 epochs of training|
-------------------------------------------------------------------------------

Training loss:		[1.0867689, 0.6528642, 0.5661395, 0.52863705, 0.51213443, 0.5073384, 0.46672076, 0.4546424, 0.4447795, 0.44817638, 0.4436681, 0.42907992, 0.42442623, 0.40833628, 0.41765544, 0.40699196, 0.3944471, 0.40329665, 0.39697438, 0.38386303]

Validation loss:	[0.6924095, 0.56056184, 0.52611995, 0.5089408, 0.4943622, 0.48108596, 0.47100875, 0.45790562, 0.4462174, 0.44172728, 0.43143767, 0.41881633, 0.41113362, 0.40643442, 0.39944822, 0.39482912, 0.3904964, 0.38050324, 0.37635624, 0.37209097]

Lowest Training loss:	0.38386303186416626

Lowest Validation loss:	0.3720909655094147

Final eval data:

IoU metric: bbox
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.145
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 

## Run inference on random image from test set using trained model to create detections
- Ground truth annotation
- Raw detection
- Confidence Score threshold of 0.5
- save loss graph in `artifacts/` under `/project_name/`

In [10]:
!python3 detections.py -d 'artifacts' -n 'planes_sgd'


-------------------------------------------------------------------------------
Detections with |FasterRCNN_ResNet50: 20 epochs of training|
-------------------------------------------------------------------------------

-------------------------------------------------------------------------------
Saving detection artifacts to: artifacts/planes_sgd/analysis/
-------------------------------------------------------------------------------

